net/httpbody.go

```diff
@@ -0,0 +1,270 @@
+		input:  rc,
+		log:    log,
+	}
+}
+
+func (lb *logBody) Read(p []byte) (int, error) {
+	n, err := lb.input.Read(p)
+	if n > 0 {
+		lb.log("%s%s", lb.fmtstr, p)
+	}
+	return n, err
+}
+
+func (lb *logBody) Close() error {
+	return lb.input.Close()
+}
+
+func LogBody(ctx context.Context, fmtstr string, log func(format string, args ...interface{}), rc io.ReadCloser) io.ReadCloser {
+	return newLogBody(ctx, fmtstr, log, rc)
+}
+
+type matcher struct {
+	ctx               context.Context
+	once              sync.Once
+	input             io.ReadCloser
+	f                 func([]byte) (int, error)
+	maxBufferSize     uint64
+	maxBufferHandling MaxBufferHandling
+	readBuffer        []byte
+
+	ready   *bytes.Buffer
+	pending *bytes.Buffer
+
+	metrics metrics.Metrics
+
+	err    error
+	closed bool
+}
+
+var (
+	ErrMatcherBufferFull = errors.New("matcher buffer full")
+)
+
+func newMatcher(
+	ctx context.Context,
+	input io.ReadCloser,
+	f func([]byte) (int, error),
+	maxBufferSize uint64,
+	mbh MaxBufferHandling,
+) *matcher {
+
+	rsize := defaultReadBufferSize
+	if maxBufferSize < rsize {
+		rsize = maxBufferSize
+	}
+
+	return &matcher{
+		ctx:               ctx,
+		once:              sync.Once{},
+		input:             input,
+		f:                 f,
+		maxBufferSize:     maxBufferSize,
+		maxBufferHandling: mbh,
+		readBuffer:        make([]byte, rsize),
+		pending:           bytes.NewBuffer(nil),
+		ready:             bytes.NewBuffer(nil),
+		metrics:           metrics.Default,
+	}
+}
+
+func (m *matcher) readNTimes(times int) (bool, error) {
+	var consumedInput bool
+	for i := 0; i < times; i++ {
+		n, err := m.input.Read(m.readBuffer)
+		_, err2 := m.pending.Write(m.readBuffer[:n])
+
+		if n > 0 {
+			consumedInput = true
+		}
+
+		if err != nil {
+			return consumedInput, err
+		}
+		if err2 != nil {
+			return consumedInput, err2
+		}
+
+	}
+
+	return consumedInput, nil
+}
+
+func (m *matcher) fill(requested int) error {
+	readSize := 1
+	for m.ready.Len() < requested {
+		consumedInput, err := m.readNTimes(readSize)
+		if !consumedInput {
+			io.CopyBuffer(m.ready, m.pending, m.readBuffer)
+			return err
+		}
+
+		if uint64(m.pending.Len()) > m.maxBufferSize {
+			switch m.maxBufferHandling {
+			case MaxBufferAbort:
+				return ErrMatcherBufferFull
+			default:
+				select {
+				case <-m.ctx.Done():
+					m.Close()
+					return m.ctx.Err()
+				default:
+				}
+				_, err := m.f(m.pending.Bytes())
+				if err != nil {
+					return err
+				}
+				m.pending.Reset()
+				readSize = 1
+			}
+		}
+
+		readSize *= 2
+	}
+	return nil
+}
+
+func (m *matcher) Read(p []byte) (int, error) {
+	if m.closed {
+		return 0, ErrClosed
+	}
+
+	if m.ready.Len() == 0 && m.err != nil {
+		return 0, m.err
+	}
+
+	if m.ready.Len() < len(p) {
+		m.err = m.fill(len(p))
+	}
+
+	if m.err == ErrMatcherBufferFull {
+		return 0, ErrMatcherBufferFull
+	}
+
+	if m.err == ErrBlocked {
+		m.metrics.IncCounter("blocked.requests")
+		return 0, ErrBlocked
+	}
```

Comment:
metrics yes, but I am not sure if we can't really move the if block because we need to stop Read() to block the body streaming to the filter.
I will try to move metrics and maybe the error handling to the filter.
I guess an err handling `switch` would be also nicer.

Reply:
done